documents/entity/model_test.go

```diff
@@ -0,0 +1,382 @@
+	assert.Nil(t, err, "Init must pass")
+	assert.Nil(t, m.EntitySalts, "salts must be nil")
+
+	dr, err := m.CalculateDataRoot()
+	assert.Nil(t, err, "calculate must pass")
+	assert.False(t, utils.IsEmptyByteSlice(dr))
+	assert.NotNil(t, m.EntitySalts, "salts must be created")
+}
+
+func TestEntity_CreateProofs(t *testing.T) {
+	i := createEntity(t)
+	rk := i.Document.Roles[0].RoleKey
+	pf := fmt.Sprintf(documents.CDTreePrefix+".roles[%s].collaborators[0]", hexutil.Encode(rk))
+	proof, err := i.CreateProofs([]string{"entity.legal_name", pf, documents.CDTreePrefix + ".document_type"})
+	assert.Nil(t, err)
+	assert.NotNil(t, proof)
+	tree, err := i.CoreDocument.DocumentRootTree()
+	assert.NoError(t, err)
+
+	// Validate entity_number
+	valid, err := tree.ValidateProof(proof[0])
+	assert.Nil(t, err)
+	assert.True(t, valid)
+
+	// Validate roles
+	valid, err = tree.ValidateProof(proof[1])
+	assert.Nil(t, err)
+	assert.True(t, valid)
+
+	// Validate []byte value
+	acc := identity.NewDIDFromBytes(proof[1].Value)
+	assert.True(t, i.AccountCanRead(acc))
+
+	// Validate document_type
+	valid, err = tree.ValidateProof(proof[2])
+	assert.Nil(t, err)
+	assert.True(t, valid)
+}
+
+func createEntity(t *testing.T) *Entity {
+	i := new(Entity)
+	err := i.InitEntityInput(testingdocuments.CreateEntityPayload(), defaultDID.String())
+	assert.NoError(t, err)
+	_, err = i.CalculateDataRoot()
+	assert.NoError(t, err)
+	_, err = i.CalculateSigningRoot()
+	assert.NoError(t, err)
+	_, err = i.CalculateDocumentRoot()
+	assert.NoError(t, err)
+	return i
+}
+
+func TestEntityModel_createProofsFieldDoesNotExist(t *testing.T) {
+	i := createEntity(t)
+	_, err := i.CreateProofs([]string{"nonexisting"})
+	assert.NotNil(t, err)
+}
+
+func TestEntityModel_GetDocumentID(t *testing.T) {
+	i := createEntity(t)
+	assert.Equal(t, i.CoreDocument.ID(), i.ID())
+}
+
+func TestEntityModel_getDocumentDataTree(t *testing.T) {
+	i := Entity{LegalName: "test company"}
+	tree, err := i.getDocumentDataTree()
+	assert.Nil(t, err, "tree should be generated without error")
+	_, leaf := tree.GetLeafByProperty("entity.legal_name")
+	assert.NotNil(t, leaf)
+	assert.Equal(t, "entity.legal_name", leaf.Property.ReadableName())
+}
+
+func TestEntity_CollaboratorCanUpdate(t *testing.T) {
+	entity := createEntity(t)
+	id1 := defaultDID
+	id2 := testingidentity.GenerateRandomDID()
+	id3 := testingidentity.GenerateRandomDID()
+
+	// wrong type
+	err := entity.CollaboratorCanUpdate(new(mockModel), id1)
+	assert.Error(t, err)
+	assert.True(t, errors.IsOfType(documents.ErrDocumentInvalidType, err))
+	assert.NoError(t, testRepo().Create(id1[:], entity.CurrentVersion(), entity))
+
+	// update the document
+	model, err := testRepo().Get(id1[:], entity.CurrentVersion())
+	assert.NoError(t, err)
+	oldEntity := model.(*Entity)
+	data := oldEntity.getClientData()
+	data.LegalName = "new legal name"
+	err = entity.PrepareNewVersion(entity, data, []string{id3.String()})
```

Comment:
@vedhavyas I am adding `id3` as a collaborator and by default, he has write-access to everything?

Reply:
yeap